bus-mapping/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+//! Crate docs
+
+#![cfg_attr(docsrs, feature(doc_cfg))]
+// Temporary until we have more of the crate implemented.
+#![allow(dead_code)]
+// Catch documentation errors caused by code changes.
+#![deny(broken_intra_doc_links)]
+#![deny(missing_debug_implementations)]
+#![deny(missing_docs)]
+#![deny(unsafe_code)]
+
+use std::{
+    collections::BTreeMap,
+    iter::Iterator,
+    ops::{Index, IndexMut},
+};
+
+use itertools::Itertools;
+use pasta_curves::arithmetic::FieldExt;
+
+// -------- EVM Circuit
+// gc as index of each row corresponds to the state after the opcode is performed
+// 1 gc has 1 entry
+//`StackElem{key 	val 	stack_op(READ/WRITE) 	gc   OPCODE    CallID(later_future)}`
+
+//------ State Circuit
+// Group target (stack or memory)
+// Group by key (location in memory/storage)
+// Sorty by gc
+//`MemoryElem{target 	gc 	val1 	val2 	val3}`
+
+#[derive(Debug, Clone)]
+struct BlockConstants<F: FieldExt> {
+    hash: [u8; 256], // Until we know how to deal with it
+    coinbase: F,
+    timestamp: F,
+    number: F,
+    difficulty: F,
+    gaslimit: F,
+    chain_id: F,
+}
+
+#[derive(Debug, Clone)]
+enum RW {
+    READ,
+    WRITE,
+}
+
+/// Doc
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Target {
+    /// Doc
+    Memory,
+    /// Doc
+    Stack,
+    /// Doc
+    Storage,
+}
+
+/// Doc
+#[derive(Debug, Clone)]
+pub struct Operation<F: FieldExt> {
+    mem_op: RW,
+    target: Target,
+    key: usize,
+    value: F,
+    opcode_info: String,
```

Comment:
Does opcode_info hold strings like "ADD"? Say if we want to run an evm "ADD", so we have 2 stack reads and 1 stack write. And we end up have 3 operations?